test/factories/cluster_group.rb

```diff
@@ -6,48 +6,48 @@
     name { study_file.name }
     study { study_file.study }
 
-    # create a cluster group with data arrays populated according to a passed-in cell_data object
+    # create a cluster group with data arrays populated according to a passed-in cell_input object
     factory :cluster_group_with_cells do
       transient do
-        # cell_data is a hash of three (or 4) arrays: cells, x and y and z
+        # cell_input is a hash of three (or 4) arrays: cells, x and y and z
         # {
         #   x: [1, 2, 3],
         #   y: [1, 2, 3],
         #   cells: ['cellA', 'cellB', 'cellC']
         # }
-        cell_data {
+        cell_input {
           {}
         }
-        # annotations is an array of objects specifying name, type, and values for annotations
+        # annotation_input is an array of objects specifying name, type, and values for annotations
         # values should be an array in the same length and order as the 'cells' array above
         # e.g. [{ name: 'category', type: 'group', values: ['foo', 'foo', 'bar'] }]
-        annotations {
+        annotation_input {
           []
         }
       end
       cell_annotations {
-        annotations.map { |a| { name: a[:name], type: a[:type], values: a[:values].uniq } }
+        annotation_input.map { |a| { name: a[:name], type: a[:type], values: a[:values].uniq } }
       }
 
       after(:create) do |cluster, evaluator|
         [
-          {name: :x, type: 'coordinates'},
-          {name: :y, type: 'coordinates'},
-          {name: :z, type: 'coordinates'},
-          {name: :cells, type: 'cells'}
+          {name: :x, type: 'coordinates', array_name: 'x'},
+          {name: :y, type: 'coordinates', array_name: 'y'},
+          {name: :z, type: 'coordinates', array_name: 'z'},
+          {name: :cells, type: 'cells', array_name: 'text'}
         ].each do |input_type|
-          if evaluator.cell_data[input_type[:name]]
+          if evaluator.cell_input[input_type[:name]]
             FactoryBot.create(:data_array,
```

Comment:
This works nicely, but there is a problem in that you can define a cluster with `x`, `y`, and `z` coordinates but it will still be marked as `2d`.  We probably need to parameterize that, or make it auto-detect and set that attribute.

Reply:
yup, that's how it's set up now,
`cluster_type { '2d' }` defines the default for cluster_type as `'2d'`, but if you specify `cluster_type` it will override the default.

Reply:
Ah, understood - I wasn't sure you could specify that dynamically.